ui/js/components/auth.js

```diff
@@ -0,0 +1,165 @@
+"use strict";
+
+/**
+ * This component handles logging in to Taskcluster Authentication
+ *
+ * See: https://docs.taskcluster.net/manual/3rdparty
+ *
+ * This communicates to the rest of Treeherder by calling the onUserChange
+ * function on the HTML element, which in turn stores that user in $rootScope.
+ */
+treeherder.component("login", {
+    template: `
+        <span class="dropdown"
+              ng-if="user.loggedin">
+          <button id="logoutLabel" title="Logged in as: {{user.email}}" role="button"
+                  data-toggle="dropdown" data-target="#"
+                  class="btn btn-view-nav btn-right-navbar nav-persona-btn">
+            <div class="nav-user-icon">
+              <span class="fa fa-user pull-left"></span>
+            </div>
+            <span class="fa fa-angle-down lightgray"></span>
+          </button>
+          <ul class="dropdown-menu" role="menu" aria-labelledby="logoutLabel">
+              <li>
+                <a ng-click="logout()">Logout</a>
+              </li>
+          </ul>
+        </span>
+
+        <a class="btn btn-view-nav btn-right-navbar nav-login-btn"
+           ng-if="!user.loggedin"
+           ng-click="login()">Login/Register</a>
+    `,
+    bindings: {
+        onUserChange: "&"
+    },
+    controller: ['$scope', '$location', '$window', 'localStorageService',
+        'ThUserModel', '$http', 'thUrl', '$timeout',
+        function ($scope, $location, $window, localStorageService,
+                  ThUserModel, $http, thUrl, $timeout) {
```

Comment:
I took a look at your approach of ``ctrl = this`` and that seems reasonable enough.  I was using ``$scope`` so I could access the ``this`` type items in child functions and it seemed pretty familiar.  Since it's an isolate scope I didn't see it as much different.  But ``ctrl`` works well, too.  And it's more in line with the ``component`` pattern.   I had considered moving the anonymous functions out to be declared on ``this`` as well, but it just seemed like that would add more lines of code unnecessarily.  This feels like a good compromise.